src/System.Private.CoreLib/shared/System/Number.Formatting.cs

```diff
@@ -1929,6 +1927,9 @@ internal static unsafe void NumberToStringFormat(ref ValueStringBuilder sb, ref
                     }
                 }
             }
+
+            if (number.sign && (section == 0) && (sb.Length > 0))
```

Comment:
We could keep the original path and have the code down here only for the case where `scale == 0`.

For the `scale == 0` case, I didn't see a clear way to check if a digit would be printed outside duplicating the base check from the while loop.